### PR TITLE
PR from fork/1167

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR is to merge a branch on a fork of our repo into develop. 
- The orginal PR is https://github.com/LLNL/RAJA/pull/1167
- Please review and approve this one.

### CODE REVIEW

**Review by MrBurmark — APPROVED** [review]

**Review by mdavis36 — DISMISSED** [review]

**Review by homerdin — APPROVED** [review]

File: include/RAJA/policy/sycl/kernel/SyclKernel.hpp [hunk 1]
```diff
@@ -186,7 +161,7 @@ struct SyclLaunchHelperNonTrivial<sycl_launch<async0>,StmtList,Data,Types>
  * determined at runtime using the SYCL occupancy calculator.
  */
 template<bool async0, typename StmtList, typename Data, typename Types>
-struct SyclLaunchHelper<sycl_launch<async0>,StmtList,Data,Types>
+struct SyclLaunchHelper<true,sycl_launch<async0>,StmtList,Data,Types>
```
Comment (rchen20) [inline include/RAJA/policy/sycl/kernel/SyclKernel.hpp:164]: This launcher will never `wait()` on line 189 below?

Comment (rchen20) [reply include/RAJA/policy/sycl/kernel/SyclKernel.hpp:164]: Nevermind, this `true` is referring to the `bool` on line 111, which I believe is representing trivially copyable?

Comment (homerdin) [reply include/RAJA/policy/sycl/kernel/SyclKernel.hpp:164]: Yes.  We are doing this partial specialization to dispatch to either the non_trivial or trivial version of the forall.

**Review by trws — DISMISSED** [review]
Looks good to me, I'd say try to use the `*_t` templates where possible but it's all good here.

**Comment by trws** [comment]
@homerdin, @davidbeckingsale, I think we’re running into the sycl docker container lagging behind what’s expected in the code.  At least it looks like the updates to the atomic_ref are causing failures in the sycl docker test.

**Comment by rhornung67** [comment]
@trws that was my initial thought. However, the azure sycl build passed before I merged the camp bump PR. Why would that cause this failure?

**Comment by trws** [comment]
The only sycl-related thing in there was I made sure that sycl is enabled for camp when it's enabled for RAJA, but if that wasn't working before the sycl code wasn't building at all.  I suppose that's possible, but I rather hope not.

**Comment by rhornung67** [comment]
That it wasn't building before must be the case.  Grrrfff.  It looks like were using oneapi-2021.2.0 which would explain why the compiler version is too old.

**Comment by trws** [comment]
Looking back at the last build, it did build `MemUtils-SYCL` but *all* of the tests are the `Sequential` variant.  None at all that touch sycl.  No clue _why_ that's the case, but... yeah, it looks like we've been building sycl tests in sequential mode.

**Comment by rchen20** [comment]
> Looking back at the last build, it did build `MemUtils-SYCL` but _all_ of the tests are the `Sequential` variant. None at all that touch sycl. No clue _why_ that's the case, but... yeah, it looks like we've been building sycl tests in sequential mode.

@trws The SYCL atomic-ref calls are protected with the `SYCL_DEVICE` macro, which is never activated. This is probably because our Docker image doesn't see a GPU so it wouldn't work anyway.

**Comment by trws** [comment]
It’s supposed to be compiling it for offload though, even if no GPU is present it should still generate code for the target don’t you think?

**Comment by rchen20** [comment]
If I recall correctly, when I turned the `SYCL_DEVICE` flag on a while ago, it still couldn't build the atomic refs in our Docker image. Nor could I get them to build on one of our systems. Perhaps we need a more up-to-date OneAPI compiler? How does @homerdin get the SYCL device code to build & run?

**Comment by davidbeckingsale** [comment]
Hopefully this fixes the CI.

**Comment by rhornung67** [comment]
@homerdin @trws @rchen20 @mdavis36 @artv3 @MrBurmark we're back up and running with sycl builds on azure -- thanks @davidbeckingsale !. Unless, someone has changes to suggest, this should be good to merge. Please review and comment/approve. Thank you.

**Review by artv3 — APPROVED** [review]

**Review by mdavis36 — APPROVED** [review]

**Comment by trws** [comment]
Excellent, thanks @davidbeckingsale!  Merged.

**Comment by rhornung67** [comment]
Now that this is merged and running through CI again, I will enable some basic testing of SYCL in our CI so we can build on that and fill the rest in.  Look for a PR soon-ish.

**Pull request merged**